src/shred_network/duplicate_shred_listener.zig

```diff
@@ -0,0 +1,794 @@
+const DuplicateShredHandler = struct {
+    allocator: Allocator,
+    logger: Logger,
+    params: Params,
+    // Because we use UDP for packet transfer, we can normally only send ~1500 bytes
+    // in each packet. We send both shreds and meta data in duplicate shred proof, and
+    // each shred is normally 1 packet(1500 bytes), so the whole proof is larger than
+    // 1 packet and it needs to be cut down as chunks for transfer. So we need to piece
+    // together the chunks into the original proof before anything useful is done.
+    dup_buffer: std.AutoHashMapUnmanaged(Key, BufferEntry),
+    consumed: std.AutoHashMapUnmanaged(Slot, bool),
+    last_root: Slot,
+    cached_slots_in_epoch: u64,
+
+    pub fn init(allocator: Allocator, logger: Logger, params: Params) DuplicateShredHandler {
+        return .{
+            .allocator = allocator,
+            .logger = logger,
+            .params = params,
+            .dup_buffer = .empty,
+            .consumed = .empty,
+            .last_root = 0,
+            .cached_slots_in_epoch = params.epoch_schedule.slots_per_epoch,
+        };
+    }
+
+    pub fn deinit(self: *DuplicateShredHandler) void {
+        var it = self.dup_buffer.valueIterator();
+        while (it.next()) |entry| {
+            for (entry.chunks) |maybe_chunk| if (maybe_chunk) |chunk| self.allocator.free(chunk);
+        }
+        self.dup_buffer.deinit(self.allocator);
+        self.consumed.deinit(self.allocator);
+    }
+
+    pub fn handle(self: *DuplicateShredHandler, dup_shred_data: DuplicateShredData) !void {
+        self.cacheRootInfo();
+        self.maybePruneBuffer();
+        try self.handleShredData(dup_shred_data);
+    }
+
+    fn cacheRootInfo(self: *DuplicateShredHandler) void {
+        const new_root = self.params.ledger_reader.maxRoot();
+        if (new_root == self.last_root) return;
+        self.last_root = new_root;
+        const epoch = self.params.epoch_schedule.getEpoch(self.last_root);
+        self.cached_slots_in_epoch = self.params.epoch_schedule.getSlotsInEpoch(epoch);
+    }
+
+    fn shouldConsumeSlot(self: *DuplicateShredHandler, slot: Slot) !bool {
+        const max_slot = self.last_root +| self.cached_slots_in_epoch;
+        if (!(slot > self.last_root and slot < max_slot)) return false;
+        // Returns false if a duplicate proof is already ingested for the slot,
+        // and updates local `consumed` cache with blockstore.
+        const gop = try self.consumed.getOrPut(self.allocator, slot);
+        if (!gop.found_existing) {
+            gop.value_ptr.* = try self.params.ledger_reader.isDuplicateSlot(slot);
+        }
+        return !gop.value_ptr.*;
+    }
+
+    fn maybePruneBuffer(self: *DuplicateShredHandler) void {
+        if (self.dup_buffer.count() < BUFFER_CAPACITY * 2) return;
+
+        var counts: std.AutoHashMapUnmanaged(Pubkey, usize) = .empty;
+        defer counts.deinit(self.allocator);
+        var keys_to_remove = std.ArrayListUnmanaged(Key){};
```

Comment:
remember, the default field init will be removed from arraylist, use the `.empty` decl.